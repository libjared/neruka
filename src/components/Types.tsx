type SignedDuration = Duration & {
  negative: boolean;
};

type BaseTenDigit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";

type FriendlyDuration = {
  negative: boolean;
  hourTens?: BaseTenDigit;
  hourOnes?: BaseTenDigit;
  minuteTens?: BaseTenDigit;
  minuteOnes?: BaseTenDigit;
  secondTens?: BaseTenDigit;
  secondOnes: BaseTenDigit; // the only required digit
};

export type { SignedDuration, BaseTenDigit, FriendlyDuration };
