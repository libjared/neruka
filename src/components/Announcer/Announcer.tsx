import React from "react";
import { useEffect, useRef, useState } from "react";
import audioClips, { AudioClip } from "../../data/AudioClips";
import { SignedDuration } from "../Types";

type MilestoneDefinition = {
  at: SignedDuration;
  sound: AudioClip;
};

// list of milestones and their associated clips.
// sorted by `at` in descending order.
const MilestoneList: ReadonlyArray<MilestoneDefinition> = [
  {
    at: { negative: false, hours: 3 },
    sound: audioClips.vox03htogo,
  },
  {
    at: { negative: false, hours: 2 },
    sound: audioClips.vox02htogo,
  },
  {
    at: { negative: false, hours: 1 },
    sound: audioClips.vox01htogo,
  },
  {
    at: { negative: false, minutes: 30 },
    sound: audioClips.vox30mtogo,
  },
  {
    at: { negative: false, minutes: 15 },
    sound: audioClips.vox15mtogo,
  },
  {
    at: { negative: false, minutes: 14 },
    sound: audioClips.vox14,
  },
  {
    at: { negative: false, minutes: 13 },
    sound: audioClips.vox13,
  },
  {
    at: { negative: false, minutes: 12 },
    sound: audioClips.vox12,
  },
  {
    at: { negative: false, minutes: 11 },
    sound: audioClips.vox11,
  },
  {
    at: { negative: false, minutes: 10 },
    sound: audioClips.vox10,
  },
  {
    at: { negative: false, minutes: 9 },
    sound: audioClips.vox09,
  },
  {
    at: { negative: false, minutes: 8 },
    sound: audioClips.vox08,
  },
  {
    at: { negative: false, minutes: 7 },
    sound: audioClips.vox07,
  },
  {
    at: { negative: false, minutes: 6 },
    sound: audioClips.vox06,
  },
  {
    at: { negative: false, minutes: 5 },
    sound: audioClips.vox05,
  },
  {
    at: { negative: false, minutes: 4 },
    sound: audioClips.vox04,
  },
  {
    at: { negative: false, minutes: 3 },
    sound: audioClips.vox03,
  },
  {
    at: { negative: false, minutes: 2 },
    sound: audioClips.vox02,
  },
  {
    at: { negative: false, minutes: 1 },
    sound: audioClips.vox01mrem,
  },
  {
    at: { negative: false, seconds: 30 },
    sound: audioClips.vox30s,
  },
  {
    at: { negative: false, seconds: 10 },
    sound: audioClips.vox10,
  },
  {
    at: { negative: false, seconds: 9 },
    sound: audioClips.vox09,
  },
  {
    at: { negative: false, seconds: 8 },
    sound: audioClips.vox08,
  },
  {
    at: { negative: false, seconds: 7 },
    sound: audioClips.vox07,
  },
  {
    at: { negative: false, seconds: 6 },
    sound: audioClips.vox06,
  },
  {
    at: { negative: false, seconds: 5 },
    sound: audioClips.vox05,
  },
  {
    at: { negative: false, seconds: 4 },
    sound: audioClips.vox04,
  },
  {
    at: { negative: false, seconds: 3 },
    sound: audioClips.vox03,
  },
  {
    at: { negative: false, seconds: 2 },
    sound: audioClips.vox02,
  },
  {
    at: { negative: false, seconds: 1 },
    sound: audioClips.vox01,
  },
  {
    at: { negative: false, seconds: 0 },
    sound: audioClips.vox00,
  },
  {
    at: { negative: true, seconds: 10 },
    sound: audioClips.voxTimeAway,
  },
];

type AnnouncerProps = {
  duration: SignedDuration;
};

const Announcer = React.memo(({ duration }: AnnouncerProps) => {
  const audioRef = useRef<HTMLAudioElement>(null);
  // the idx of the next upcoming milestone, or the one whose sound we're currently playing.
  const [currentIdx, setCurrentIdx] = useState<number>(() =>
    findInitialIdx(duration)
  );
  // flag indicating whether the audio at `src` is loaded enough to play right now.
  const [isLoaded, setIsLoaded] = useState<boolean>(false);
  // flag indicating whether the audio is currently busy saying the milestone at currentIdx.
  const [isPlaying, setIsPlaying] = useState<boolean>(false);

  const currentMilestone: MilestoneDefinition | null =
    currentIdx < MilestoneList.length ? MilestoneList[currentIdx] : null;

  const isPassed =
    currentMilestone !== null &&
    compareDuration(duration, currentMilestone.at) !== 1;

  // TODO: what are the implications of having this inside the effect vs outside?
  const shouldPlay = isPassed && isLoaded && !isPlaying;

  useEffect(() => {
    if (!shouldPlay) {
      return;
    }
    setIsPlaying(true);
    // audioRef must exist if isLoaded is true
    audioRef.current!.play();
  }, [currentMilestone, shouldPlay]);

  return (
    <>
      <audio
        title="announcer"
        ref={audioRef}
        src={currentMilestone?.sound.src}
        onCanPlayThrough={() => {
          setIsLoaded(true);
        }}
        onEnded={() => {
          // FIXME: If the milestone that just ended has the same src as the
          // next, <audio src> attribute won't change, thus onCanPlayThrough
          // won't fire next time. We'll be waiting forever for it to load,
          // not aware it already has. We can't reliably check this, so two
          // milestones sharing the same sound src is unsupported.
          setIsLoaded(false);
          setIsPlaying(false);
          setCurrentIdx((cidx) => cidx + 1);
        }}
      />
    </>
  );
}, propsAreEqual);

function propsAreEqual(
  prevProps: Readonly<AnnouncerProps>,
  nextProps: Readonly<AnnouncerProps>
): boolean {
  return signedDurationEqual(prevProps.duration, nextProps.duration);
}

function signedDurationEqual(a: SignedDuration, b: SignedDuration): boolean {
  return (
    a.hours === b.hours &&
    a.minutes === b.minutes &&
    a.seconds === b.seconds &&
    a.negative === b.negative
  );
}

/**
 * Finds the first milestone that is less than or equal to the passed duration, and returns its index.
 * @param duration
 * @returns the index of the initial duration, or a number equal to MilestoneList.length if none were found.
 */
function findInitialIdx(duration: SignedDuration): number {
  let idx;
  for (idx = 0; idx < MilestoneList.length; idx++) {
    const milestone = MilestoneList[idx];
    if (compareDuration(duration, milestone.at) !== -1) {
      return idx;
    }
  }
  return idx; // MilestoneList.length
}

// returns -1 if left is less than right.
// returns 1 if left is greater than right.
// returns 0 if equal.
function compareDuration(left: SignedDuration, right: SignedDuration): number {
  if (left.negative && !right.negative) {
    return -1;
  }
  if (!left.negative && right.negative) {
    return 1;
  }
  const flipResult = left.negative && right.negative ? -1 : 1;
  const hrDiff = (left.hours ?? 0) - (right.hours ?? 0);
  const minDiff = (left.minutes ?? 0) - (right.minutes ?? 0);
  const secDiff = (left.seconds ?? 0) - (right.seconds ?? 0);
  if (hrDiff !== 0) {
    return Math.sign(hrDiff) * flipResult;
  }
  if (minDiff !== 0) {
    return Math.sign(minDiff) * flipResult;
  }
  if (secDiff !== 0) {
    return Math.sign(secDiff) * flipResult;
  }
  return 0;
}

export default Announcer;
export { compareDuration, findInitialIdx };
