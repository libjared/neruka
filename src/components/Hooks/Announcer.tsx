import React from "react";
import { useEffect, useRef, useState } from "react";
import audioClips, { AudioClip } from "../../data/AudioClips";
import { SignedDuration } from "../Types";

type MilestoneDefinition = {
  at: SignedDuration;
  sound: AudioClip;
};

// list of milestones and their associated clips.
// sorted by `at` in descending order.
const MilestoneList: ReadonlyArray<MilestoneDefinition> = [
  { at: { negative: false, minutes: 15 }, sound: audioClips.remain15m },
  {
    at: { negative: false, minutes: 14, seconds: 50 },
    sound: audioClips.remain15m,
  },
  { at: { negative: false, minutes: 14 }, sound: audioClips.remain15m },
  { at: { negative: false, seconds: 1 }, sound: audioClips.remain15m },
  { at: { negative: false, seconds: 0 }, sound: audioClips.remain15m },
];

type AnnouncerProps = {
  duration: SignedDuration;
};

const Announcer = React.memo(({ duration }: AnnouncerProps) => {
  const audioRef = useRef<HTMLAudioElement>(null);
  // the idx of the next upcoming milestone, or the one whose sound we're currently playing.
  const [currentIdx, setCurrentIdx] = useState<number>(() =>
    findInitialIdx(duration)
  );
  // flag indicating whether the audio at `src` is loaded enough to play right now.
  const [isLoaded, setIsLoaded] = useState<boolean>(false);
  // flag indicating whether the audio is currently busy saying the milestone at currentIdx.
  const [isPlaying, setIsPlaying] = useState<boolean>(false);

  const currentMilestone: MilestoneDefinition | null =
    currentIdx < MilestoneList.length ? MilestoneList[currentIdx] : null;

  const isPassed =
    currentMilestone !== null
      ? compareDuration(duration, currentMilestone.at) !== 1
      : false;

  // TODO: what are the implications of having this inside the effect vs outside?
  const shouldPlay = isPassed && isLoaded && !isPlaying;

  useEffect(() => {
    if (!shouldPlay) {
      return;
    }
    if (audioRef.current === null) {
      throw new Error("Expected audioRef.current not to be null.");
      // because how would this ref not be defined, if isLoaded is true?
    }
    if (currentMilestone === null) {
      throw new Error("Expected currentMilestone not to be null.");
    }
    const actualSrc = audioRef.current.src;
    const expectSrc = currentMilestone.sound.src;
    if (!actualSrc.endsWith(expectSrc)) {
      throw new Error(
        `Expected audioRef.current.src to end with ${expectSrc}, but it was ${actualSrc}.`
      );
    }
    setIsPlaying(true);
    audioRef.current.play();
  }, [currentMilestone, shouldPlay]);

  return (
    <>
      <audio
        ref={audioRef}
        src={currentMilestone?.sound.src}
        onCanPlayThrough={() => {
          setIsLoaded(true);
        }}
        onEnded={() => {
          if (audioRef.current === null) {
            throw new Error("Expected audioRef.current not to be null.");
          }
          const newSrc = currentMilestone?.sound.src;
          const currentSrc = audioRef.current.src;

          // If the milestone that just ended has the same src as the next,
          // <audio src> attribute won't change, thus onCanPlayThrough won't fire next time.
          // We'll be waiting forever for it to load, not aware it already has. So let's keep it loaded.
          if (newSrc === undefined || !isSrcEqual(currentSrc, newSrc)) {
            setIsLoaded(false);
          }
          setIsPlaying(false);
          setCurrentIdx((cidx) => cidx + 1);
        }}
      />
    </>
  );
}, propsAreEqual);

function isSrcEqual(currentSrc: string, newSrc: string): boolean {
  return currentSrc.endsWith(newSrc);
}

function propsAreEqual(
  prevProps: Readonly<AnnouncerProps>,
  nextProps: Readonly<AnnouncerProps>
): boolean {
  return signedDurationEqual(prevProps.duration, nextProps.duration);
}

function signedDurationEqual(a: SignedDuration, b: SignedDuration): boolean {
  return (
    a.hours === b.hours &&
    a.minutes === b.minutes &&
    a.seconds === b.seconds &&
    a.negative === b.negative
  );
}

/**
 * Finds the first milestone that is less than or equal to the passed duration, and returns its index.
 * @param duration
 * @returns the index of the initial duration, or a number equal to MilestoneList.length if none were found.
 */
function findInitialIdx(duration: SignedDuration): number {
  let idx;
  for (idx = 0; idx < MilestoneList.length; idx++) {
    const milestone = MilestoneList[idx];
    if (compareDuration(duration, milestone.at) !== -1) {
      return idx;
    }
  }
  return idx; // MilestoneList.length
}

// returns -1 if left is less than right.
// returns 1 if left is greater than right.
// returns 0 if equal.
function compareDuration(left: SignedDuration, right: SignedDuration): number {
  if (left.negative && !right.negative) {
    return -1;
  }
  if (!left.negative && right.negative) {
    return 1;
  }
  const flipResult = left.negative && right.negative ? -1 : 1;
  const hrDiff = (left.hours ?? 0) - (right.hours ?? 0);
  const minDiff = (left.minutes ?? 0) - (right.minutes ?? 0);
  const secDiff = (left.seconds ?? 0) - (right.seconds ?? 0);
  if (hrDiff !== 0) {
    return Math.sign(hrDiff) * flipResult;
  }
  if (minDiff !== 0) {
    return Math.sign(minDiff) * flipResult;
  }
  if (secDiff !== 0) {
    return Math.sign(secDiff) * flipResult;
  }
  return 0;
}

export default Announcer;
export { compareDuration, findInitialIdx };
