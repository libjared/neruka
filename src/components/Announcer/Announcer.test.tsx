import { render, RenderResult, screen } from "@testing-library/react";
import Announcer, { compareDuration, findInitialIdx } from "./Announcer";
import { act } from "react-test-renderer";
import { SignedDuration } from "../Types";

type SetupArgs = {
  duration: SignedDuration;
};

type SetupResult = RenderResult & {
  mediaPlay: jest.Mock<Promise<void>, []>;
  getAudio: () => HTMLAudioElement;
  fireCanPlayThrough: () => void;
  fireEnded: () => void;
};

function setup(args?: SetupArgs): SetupResult {
  const mediaPlay = jest.fn<Promise<void>, []>();
  window.HTMLMediaElement.prototype.play = mediaPlay;

  const duration = args?.duration ?? { negative: false, minutes: 15 };

  const utils = render(<Announcer duration={duration} />);

  const getAudio = () => screen.getByTitle("announcer") as HTMLAudioElement;
  const fireAudioEvent = (type: string) =>
    act(() => {
      // TODO: replace with fireEvent
      getAudio().dispatchEvent(new Event(type));
    });
  const fireCanPlayThrough = () => fireAudioEvent("canplaythrough");
  const fireEnded = () => fireAudioEvent("ended");

  return {
    ...utils,
    mediaPlay,
    getAudio,
    fireCanPlayThrough,
    fireEnded,
  };
}

it("matches snapshot", () => {
  const { asFragment } = setup();
  expect(asFragment()).toMatchSnapshot();
});

it("plays when loaded and passed", () => {
  const { mediaPlay, fireCanPlayThrough } = setup();
  expect(mediaPlay).not.toHaveBeenCalled();
  fireCanPlayThrough();
  expect(mediaPlay).toHaveBeenCalledTimes(1);
});

it("targets the next milestone when the sound clip ends", () => {
  const { getAudio, fireCanPlayThrough, fireEnded } = setup();
  const audio = getAudio();
  expect(audio.src).toBe("http://localhost/fifteen%20minutes%20to%20go.ogg");
  fireCanPlayThrough();
  fireEnded();
  expect(audio.src).toBe("http://localhost/fourteen.ogg");
});

it("doesn't target anything when all milestones have passed", () => {
  const { getAudio } = setup({
    duration: { negative: true, seconds: 16 },
  });
  const audio = getAudio();
  expect(audio.src).toBe("");
});

describe("compareDuration", () => {
  it("returns -1 if left < right", () => {
    expect(
      compareDuration(
        { negative: false, minutes: 14, seconds: 59 },
        { negative: false, minutes: 15 }
      )
    ).toBe(-1);
  });

  it("returns 1 if left > right", () => {
    expect(
      compareDuration(
        { negative: false, minutes: 15 },
        { negative: false, minutes: 14, seconds: 59 }
      )
    ).toBe(1);
  });

  it("returns 0 if left = equal", () => {
    expect(
      compareDuration(
        { negative: true, minutes: 14, seconds: 59 },
        { negative: true, minutes: 14, seconds: 59 }
      )
    ).toBe(0);
  });

  it("returns -1 if -left and +right", () => {
    expect(
      compareDuration(
        { negative: true, minutes: 14, seconds: 59 },
        { negative: false, minutes: 1, seconds: 1 }
      )
    ).toBe(-1);
  });

  it("returns 1 if +left and -right", () => {
    expect(
      compareDuration(
        { negative: false, minutes: 1, seconds: 1 },
        { negative: true, minutes: 14, seconds: 59 }
      )
    ).toBe(1);
  });

  it("returns -1 if both are negative and left < right", () => {
    expect(
      compareDuration(
        { negative: true, minutes: 15, seconds: 5 },
        { negative: true, minutes: 14, seconds: 5 }
      )
    ).toBe(-1);
  });

  it("returns 1 if both are negative and left > right", () => {
    expect(
      compareDuration(
        { negative: true, minutes: 14, seconds: 5 },
        { negative: true, minutes: 15, seconds: 5 }
      )
    ).toBe(1);
  });
});

describe("findInitialIdx", () => {
  it("returns 0 for a really large duration", () => {
    const duration = { negative: false, hours: 23 };
    expect(findInitialIdx(duration)).toBe(0);
  });

  it("returns the 14m milestone for exactly 14m", () => {
    const duration = { negative: false, minutes: 14, seconds: 0 };
    expect(findInitialIdx(duration)).toBe(7);
  });

  it("returns the next milestone for 13m59s", () => {
    const duration = { negative: false, minutes: 13, seconds: 59 };
    expect(findInitialIdx(duration)).toBe(8);
  });

  it("returns length for a very overdue duration", () => {
    const duration = { negative: true, hours: 23 };
    expect(findInitialIdx(duration)).toBe(34);
  });
});
