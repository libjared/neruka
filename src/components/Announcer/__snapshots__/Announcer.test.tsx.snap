// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <audio
    src="fifteen minutes to go.ogg"
    title="announcer"
  />
</DocumentFragment>
`;
